### PR TITLE
Don't fail when kernel.pid_max fails to be set (bsc#1227117)

### PR DESCRIPTION
**Pull request by Werkov**
32b kernels only allow 2^15 of pid_max at most. sysctl write fails with EINVAL on such archs. Adjust the sysctl entry so that the write is not considered a fatal error for whole systemd-sysctl.service.

### CODE REVIEW

**Comment by thkukuk** [comment]
I don't like that, since this is inconsistent behavior on different architectures and will even hide real errors on x86-64.

**Comment by Werkov** [comment]
What would you suggest for 32b instead?

The hidden failure is not serious in this case -- system still boots and the setting is not security relevant.
(Actually, the same line could be used as an argument for _visible_ failure in the original bug report :thinking: .)

**Comment by Vogtinator** [comment]
IMO it should conditionally set the highest possible value depending on arch. The makefile could do that on install for instance

**Comment by thkukuk** [comment]
Don't forget: security team stores hashes of this files, so modifying them on the fly during build time may become difficult.
I think Danilos idea of having something like `50-pidmax-32bit.conf` and installing only the right variant maybe a better solution.

**Comment by Werkov** [comment]
FTR, there is no such need on 32b, the kernel's pid_max is cranked up to the max (max max) by default. So it's sufficient to not write anything at all on those archs, IOW not installing any pid_max sysctl file there.

**Comment by Vogtinator** [comment]
> FTR, there is no such need on 32b, the kernel's pid_max is cranked up to the max (max max) by default. So it's sufficient to not write anything at all on those archs, IOW not installing any pid_max sysctl file there.

In that case I suggest just splitting that file out and only installing it on 64bit archs.

Systemd already ships this file there so can we just use that instead?

**Comment by Werkov** [comment]
> In that case I suggest just splitting that file out and only installing it on 64bit archs.

Something like I've force-pushed? (I'm not RPM-confident with the arch-names.)

> Systemd already ships this file there so can we just use that instead?

Even if one accepts the idea of setting system settings in upstream package, I still think we concentrate all various settings of our distros in `aaa_base`, so that distro policies are at one place.

**Comment by Vogtinator** [comment]
> > In that case I suggest just splitting that file out and only installing it on 64bit archs.
> 
> Something like I've force-pushed? (I'm not RPM-confident with the arch-names.)

Yep, just the Makefile should decide whether to install it IMO, otherwise `make install` has a different result than the RPM.

> > Systemd already ships this file there so can we just use that instead?
> 
> Even if one accepts the idea of setting system settings in upstream package, I still think we concentrate all various settings of our distros in `aaa_base`, so that distro policies are at one place. Cc: @fbuihuu

Upstream packages already ship sysctl files. If needed, we can deviate from that in additional files. Copying exactly what upstream does in a downstream package is IMO just duplicate work.

**Comment by fbuihuu** [comment]
> Even if one accepts the idea of setting system settings in upstream package, I still think we concentrate all various settings of our distros in `aaa_base`, so that distro policies are at one place.

That's correct. Also the idea was that these policies had to be carefully chosen/reviewed by distros instead of being silently introduced by upstream (and potentially noticed later if they introduced (perf) regressions).

**Comment by Werkov** [comment]
> otherwise make install has a different result than the RPM.

Why is that a problem?

(I considered the RPM filtering a common practice and I don't know how to easily check arch in Makefile.)

**Comment by Vogtinator** [comment]
> > Even if one accepts the idea of setting system settings in upstream package, I still think we concentrate all various settings of our distros in `aaa_base`, so that distro policies are at one place.
> 
> That's correct. Also the idea was that these policies had to be carefully chosen/reviewed by distros instead of being silently introduced by upstream (and potentially noticed later if they introduced (perf) regressions).

Should probably be part of the systemd update process then? Otherwise it'll get out of sync with distro config one way or the other.

> Why is that a problem?

If you want to test changes locally with `make install`, you end up with a different result. Currently on x86 a failed systemd-sysctl.service.

**Comment by Werkov** [comment]
Force-pushed the arch-filtering inside Makefile.

**Comment by thkukuk** [comment]
@Werkov the package is no longer building on x86:
```
[   11s] error: Installed (but unpackaged) file(s) found:
[   11s]    /usr/lib/sysctl.d/50-pid-max.conf
```

**Comment by Werkov** [comment]
The OBS repo states i586 but it's apparently built as i386:

    [   11s] Processing files: aaa_base-debuginfo-84.87+git20240723.a84995d-50.1.i586

And log header even lists more:

    [    0s] Using BUILD_ARCH=i586:i486:i386

My OBS-fu is weak for this, what should RPM_ARCH be filtered by? (And what would be the right name to fix this on 32b ARM (from the original bug report)?)

**Comment by Vogtinator** [comment]
You need to use `%ifnarch %x86 %arm`

**Comment by Vogtinator** [comment]
(In any case, I'm still against having this here in aaa_base, outside of where it's already maintained upstream)

**Comment by Werkov** [comment]
> You need to use %ifnarch %ix86 %arm

Is there an equivalent for the Makefile filtering? :-p

**Comment by Werkov** [comment]
Thanks for the patterns, those seem to work in OBS CI.

**Review by thkukuk — APPROVED** [review]

**Comment by ggardet** [comment]
> @Werkov the package is no longer building on x86:
> 
> ```
> [   11s] error: Installed (but unpackaged) file(s) found:
> [   11s]    /usr/lib/sysctl.d/50-pid-max.conf
> ```

Same problem on armv6 and armv7, see: https://build.opensuse.org/package/show/openSUSE:Factory:ARM/aaa_base

**Comment by Werkov** [comment]
I believe this
```diff
--- a/Makefile
+++ b/Makefile
@@ -9,7 +9,7 @@ all: $(bin_PROGRAMS)
 install: all
        cp -a files/* $(DESTDIR)/
        install -m755 get_kernel_version $(DESTDIR)/usr/bin
-ifneq ($(filter i%86 armv%,$(RPM_ARCH)),)
+ifneq ($(filter i%86 arm%,$(RPM_ARCH)),)
        rm -vf $(DESTDIR)/usr/lib/sysctl.d/50-pid-max.conf
 endif
 ```
could fix it, I'm waiting for some confirmations.

**Pull request merged**
